docs/design/uplink-scopes.md

```diff
@@ -0,0 +1,67 @@
+
+```go
+type BucketShare struct {
+    Bucket          string
+    EncryptedPath   string
+    UnencryptedPath string
+    EncryptionKey   []byte
+}
+```
+
+If a client attempts to interact with a bucket, and no `BucketShare` matches, and there is a project key set, a key for that bucket is derived using the bucket name and the project key.
+
+When a client attempts to interact with a bucket, it finds the longest matching unencrypted path, breaking ties by using the later one, and appends the rest of the path, encrypted with the key the bucket share specifies, to the bucket share's encrypted path. For example, if one had this set of `BucketShare` objects:
+
+```go
+paths := []BucketShare{
+    {Bucket: "x", EncryptedPath: "/e1/e2/e3", UnecryptedPath: "/a/b/c", EncryptionKey: "key1"},
+    {Bucket: "x", EncryptedPath: "/e1",       UnecryptedPath: "/a",     EncryptionKey: "key2"},
+    {Bucket: "x", EncryptedPath: "/e4",       UnecryptedPath: "/f",     EncryptionKey: "key3"},
+}
+```
+
+Then if one asked for `sj://x/a/d`, the 2st path would be the longest match and `"key2"` would be used to encrypt `d` and it would be appened to `"/e1"`. Similarly, if one asked for `sj://x/a/b/c/f`, then the 1st path would be the longest match and `"key1"` would be used to encrypt `f` and it would be appened to `"/e1/e2/e3"`. If one asked for `sj://x/f/g`, then `"key3"` would be used. Asking for `sj://x/g` would return not found.
+
+Note that the encrypted path does not need to match the unencrypted path, and this allows for having local views of bucket paths in the actual bucket.
+
+## Rationale
+
+This proposal is not a radical departure of the current method, just an extension to allowing a collection of paths to compose a bucket at any unencrypted spot. Thus, it shoudn't require any changes on the server side, reducing risk and allowing for earlier shipping. For example, any current configuration can be represented as a scope with a `ProjectEncrytionKey` specified and no bucket shares.
+
+Longest matching begs a question about consistency. In principle, it's possible to derive `"key1"` from the information provided by `"key2"`. If they disagree, an error could be thrown. Consistency does mean that you can't have differently keyed paths as part of your bucket, which may be useful. Inconsistency means that a bucket or project could contain objects that are invisible or undecryptable by clients. This design makes no requirement for consistency, and chooses to hide and warn about paths that cannot decrypt.
+
+The design does mean that if you create a new bucket on the web, you must also create the bucket in your clients, unless you have the project key. Since users are expected to manage their own encryption keys, and since creating a bucket from the command line can add the associated bucket share, this extra effort is justified.
```

Comment:
I don't really understand this paragraph. What is specific to creating a bucket on the web (I know we even don't have this at the moment)?

Reply:
You don't want to give your bucket's encryption key to the website when creating it, which means that you'd have to create the bucket in the web, then locally add the bucket with the encryption key.

I'll add a clarifying sentence.

Reply:
Oh, you mean "requesting a web server to create the bucket using server-side execution"? I believe we can cover the web use case without transferring the encryption key to the web server. We can have uplink executing in a client-side JavaScript.

Reply:
I'm just trying to point out that if you have a key per project, no client side information needs to be updated if a bucket is created on the server. If you allow buckets to have independent keys, then client information does need to be updated.